Neo4j.Driver/Neo4j.Driver/Internal/Routing/ClusterConnectionPool.cs

```diff
@@ -99,72 +93,113 @@ public Task<IConnection> AcquireAsync(Uri uri)
             return pool.AcquireAsync(ignored);
         }
 
-        // This is the ultimate method to add a pool
-        private void Add(Uri uri)
+        public void Add(IEnumerable<Uri> servers)
         {
-            _pools.GetOrAdd(uri, CreateNewConnectionPool);
+            foreach (var uri in servers)
+            {
+                _pools.AddOrUpdate(uri, CreateNewConnectionPool, ActivateConnectionPool);
+            }
             if (IsClosed)
             {
                 // Anything added after dispose should be directly cleaned.
                 Clear();
                 throw new ObjectDisposedException(GetType().Name,
-                    $"Failed to create connections with server {uri} as the driver has already started to dispose.");
+                    $"Failed to create connections with servers {servers.ToContentString()} as the driver has already started to dispose.");
             }
         }
 
-        public void Add(IEnumerable<Uri> servers)
+        public async Task AddAsync(IEnumerable<Uri> servers)
         {
             foreach (var uri in servers)
             {
-                Add(uri);
+                _pools.AddOrUpdate(uri, CreateNewConnectionPool, ActivateConnectionPool);
+            }
+            if (IsClosed)
+            {
+                // Anything added after dispose should be directly cleaned.
+                await ClearAsync().ConfigureAwait(false);
+                throw new ObjectDisposedException(GetType().Name,
+                    $"Failed to create connections with servers {servers.ToContentString()} as the driver has already started to dispose.");
             }
         }
 
-        public void Update(IEnumerable<Uri> servers)
+        public void Update(IEnumerable<Uri> added, IEnumerable<Uri> removed)
         {
-            foreach (var uri in _pools.Keys)
+            Add(added);
+            foreach (var uri in removed)
             {
-                if (!servers.Contains(uri))
+                if (_pools.TryGetValue(uri, out var pool))
                 {
-                    Purge(uri);
+                    pool.Deactivate();
+                    if (pool.NumberOfInUseConnections == 0)
+                    {
+                        Purge(uri);
+                    }
                 }
             }
-            foreach (var uri in servers)
+        }
+
+        public async Task UpdateAsync(IEnumerable<Uri> added, IEnumerable<Uri> removed)
+        {
+            await AddAsync(added).ConfigureAwait(false);
+            // TODO chain this part and use task.waitAll
```

Comment:
Is this TODO still relevant?

Reply:
This is the second TODO. Will do in another PR.